### PR TITLE
Support building wvdecrypter for iOS / tvOS

### PR DESCRIPTION
**Pull request by matthuisman**
This PR adds support for building wvdecrypter for iOS and tvOS.

It does this by targeting the native_library_ios.mm file for iOS or tvOS (darwin_embedded)
Originally this file was not used upstream by google. just stubs '"Not implemented."

I have simply copied the code from the native_library_mac.mm and removed the calls to CFBundleOpenBundleResourceMap
This is due to CFBundleOpenBundleResourceMap not being available on tvOS.
We shouldn't need this as we are only trying to open a dynlib (libwidevinecdm.dylib) which can use the dlopen code

Now, the main issue is that tvOS and iOS are locked down (like UWP)
So they won't even be able to load the widevine blob if it's downloaded on demand.
It needs to be in the signed kodi package.

The user actually needs to download the kodi deb package and sign it (or install deb via jailbreak).
So at least with these changes, they can copy the CDM into the package themselves before signing (see my comment below)

It also means we should get better error logs. 
At the moment, the log doesn't really say why a Widevine didn't play. it just doesnt find the decrypter and tries to playback.
But there is no error about no decrypter. With this PR, there will be a decrypter so at least we get a bit further and hopefully get a better error in the log when the decrypter fails to load the widevine blob

**Still not even sure if iOS / tvOS will even load the widevine blob - waiting for testings.**

### CODE REVIEW

**Comment by matthuisman** [comment]
@glennguy 
this reminds me - we should look at bringing the decrypter inside the main library.
Stops an extra dlopen() for starters :) Might speed up playback a bit.
No need to search for the decrypter, then open it which then opens the widevine libarary

**Comment by matthuisman** [comment]
example scripts for updating the tvOS / iOS debs with new IA and Widevine blob

tvOS:
```
mkdir /tmp/deb-exp && cd /tmp/deb-exp
wget -O kodi.deb https://mirrors.kodi.tv/releases/darwin/tvos/org.xbmc.kodi-tvos_19.3-0-tvos.deb?https=1
sudo dpkg-deb -R kodi.deb kodi-with-widevine/
sudo mkdir kodi-with-widevine/Applications/Kodi.app/AppData/AppHome/cdm
sudo wget -O kodi-with-widevine/Applications/Kodi.app/AppData/AppHome/cdm/libwidevinecdm.dylib https://f.mjh.nz/ia/libwidevinecdm_ios.dylib
sudo rm -r kodi-with-widevine/Applications/Kodi.app/AppData/AppHome/addons/inputstream.adaptive
sudo wget -O ia.zip https://f.mjh.nz/ia/tvos-19.0.0.zip
sudo unzip -q ia.zip -d kodi-with-widevine/Applications/Kodi.app/AppData/AppHome/addons/
sudo dpkg-deb -b kodi-with-widevine/
sudo rm -r kodi-with-widevine
sudo rm kodi.deb
sudo rm ia.zip
ls
```
iOS:
```
mkdir /tmp/deb-exp && cd /tmp/deb-exp
wget -O kodi.deb https://mirrors.kodi.tv/releases/darwin/ios-arm64/org.xbmc.kodi-ios_19.3-0_iphoneos-arm64.deb?https=1
sudo dpkg-deb -R kodi.deb kodi-with-widevine/
sudo mkdir kodi-with-widevine/Applications/Kodi.app/AppData/AppHome/cdm
sudo wget -O kodi-with-widevine/Applications/Kodi.app/AppData/AppHome/cdm/libwidevinecdm.dylib https://f.mjh.nz/ia/libwidevinecdm_ios.dylib
sudo rm -r kodi-with-widevine/Applications/Kodi.app/AppData/AppHome/addons/inputstream.adaptive
sudo wget -O ia.zip https://f.mjh.nz/ia/ios-19.0.0.zip
sudo unzip -q ia.zip -d kodi-with-widevine/Applications/Kodi.app/AppData/AppHome/addons/
sudo dpkg-deb -b kodi-with-widevine/
sudo rm -r kodi-with-widevine
sudo rm kodi.deb
sudo rm ia.zip
ls
```

**Comment by matthuisman** [comment]
as per here:
https://forum.kodi.tv/showthread.php?tid=363757&pid=3052325#pid3052325

looks like iOS / tvOS can't open the mac arm64 widevine blob anyway, so dead end.
Was worth a shot :)

@glennguy
should we still build the decrypter for ios and tvos but keep the original google native_library_ios.mm
that should raises not implemented errors.
Again, might be better than currently not finding a decrypter and trying to playback without it leading to no real error?
Basically just remove the changes I made to native_library_ios.mm and bring in the rest of the changes?

**Comment by tmm1** [comment]
It would take a bit of work but I bet you can change the mach-o LC_BUILD_VERSION and make some progress. Although it's likely the blob is also using other macOS specific libraries or functions.

**Comment by matthuisman** [comment]
thanks, as I don't have any ios devices, i'd need to leave that to someone else.

only other change that would be needed is the decrypter path would need to be updated to point to the read-only path the actual cdm is located. eg. special://xbmc/cdm

**Comment by sy6sy2** [comment]
I have both an iOS and tvOS jailbroken devices. If I can do something to help let me know 😉

**Comment by wormpy001** [comment]
I really hope you guys can fix it. Really big thanks for trying

**Comment by sy6sy2** [comment]
Hi @matthuisman, I patched the official `libwidevinecdm.dylib` Mac ARM file (version 4.10.2391.0-mac-arm64) to put iOS platform in `LC_BUILD_VERSION` section instead of Mac platform. 
But your modified version of ISA URL are dead. Any way to test my patched version on my jailbroken phone?
Thank you!

**Comment by matthuisman** [comment]
@sy6sy2 
the libwidevinecdm.dylib needs to be "baked in" anyway
Can you upload your libwidevinecdm.dylib and then i'll make some new builds with it ready to test

**Comment by sy6sy2** [comment]
[libwidevinecdm_ios.dylib.zip](https://github.com/xbmc/inputstream.adaptive/files/7549959/libwidevinecdm_ios.dylib.zip)
Thank you for your help.
Just to be honest, I have absolutely no idea if this patched version will work 😕. But, because I have a jailbroken phone, if you are able to build I minimal working version of ISA for iOS, I will be able to directly modified the libwidevinecdm.dylib file used by the add-on directly in Kodi to test if I can do some progress about that...

**Comment by matthuisman** [comment]
@sy6sy2 
ok, try this:
https://drive.google.com/file/d/1vZzwgjlPSZWn95mnLLfQOm4ytn_tN87G/view?usp=sharing
this is 19.3 for iOS with modified Inputstream Adaptive and your modified libwidevinecdm

**Comment by sy6sy2** [comment]
Thank you @matthuisman for this build test.
Unfortunately and without any surprise, some dependency libs are missing on iOS 😕:

```
2021-11-17 23:27:39.207 T:537216   DEBUG <general>: AddOnLog: inputstream.adaptive: CDM LoadNativeLibrary error: dlopen(/var/mobile/Library/Preferences/Kodi/cdm/libwidevinecdm.dylib, 1): Library not loaded: /System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices
                                                     Referenced from: /private/var/mobile/Library/Preferences/Kodi/cdm/libwidevinecdm.dylib
                                                     Reason: image not found
2021-11-17 23:27:39.207 T:537216   ERROR <general>: AddOnLog: inputstream.adaptive: Unable to load widevine shared library (/var/mobile/Library/Preferences/Kodi/cdm/libwidevinecdm.dylib)
2021-11-17 23:27:39.207 T:537216   ERROR <general>: AddOnLog: inputstream.adaptive: OpenDRMSystem failed
```

**Comment by matthuisman** [comment]
can you confirm /private/var/mobile/Library/Preferences/Kodi/cdm/libwidevinecdm.dylib is actually there?
Just checking i didn't stuff up packaging it up

**Comment by sy6sy2** [comment]
No the file was not in the correct folder. 
I had to copy it to test but I don't remember where I find it. 
Do you want the details?

**Comment by RHStonks** [comment]
@matthuisman @sy6sy2 any updates. Did you manage to get widevine working on iOS?

**Comment by steeve** [comment]
Hey folks. I see that the youtube app contains the following framework: `widevine_cdm_secured_ios.framework`, so I'm guessing there may be a way out !

**Comment by matthuisman** [comment]
you'd need to ask Inputstream Adaptive devs to add support for widevine_cdm_secured_ios.framework similar to how they use Androids build-in cdm.
